File: leetcode/2001-3000/2301-2400/2381-shifting-letters-ii/ShiftingLettersII.ts
function shiftingLetters(s: string, shifts: number[][]): string {
  const n = s.length;
  let changes = new Array(n + 1).fill(0);

  for (const [left, right, direction] of shifts) {
    const v = direction === 1 ? 1 : -1;

    changes[left] += v;
    changes[right + 1] -= v;
  }

  for (let i = 1; i <= n; i++) {
    changes[i] += changes[i - 1];
  }

  let output = "";

  for (let i = 0; i < n; i++) {
    let c = ((s.charCodeAt(i) - 97 + changes[i]) % 26 + 26) % 26 + 97
    // let c = (((s.charCodeAt(i) - 97) + changes[i]) % 26 + 26) % 26 + 97;

    output = output + String.fromCharCode(c);
  }

  return output;
};

// console.log(shiftingLetters("abc", [[0, 1, 0], [1, 2, 1], [0, 2, 1]]));
console.log(shiftingLetters("dztz", [[0, 0, 0], [1, 1, 1]]));

File: leetcode/2001-3000/2301-2400/2381-shifting-letters-ii/ShiftingLettersII.java
public class ShiftingLettersII {
  public static void main(String[] args) {
    String s = "abc";
    int[][] shifts = {
        { 0, 1, 0 },
        { 1, 2, 1 },
        { 0, 2, 1 }
    };

    System.out.println(new Solution().shiftingLetters(s, shifts));
  }
}

class Solution {
  public String shiftingLetters(String s, int[][] shifts) {
    int[] changes = new int[s.length() + 1];
    int n = s.length();

    for (int[] shift : shifts) {
      int left = shift[0];
      int right = shift[1];
      int direction = shift[2] == 1 ? 1 : -1;

      changes[left] += direction;
      changes[right + 1] -= direction;
    }

    for (int i = 1; i <= n; i++) {
      changes[i] += changes[i - 1];
    }

    StringBuilder sb = new StringBuilder();

    for (int i = 0; i < n; i++) {
      int c = ((s.charAt(i) - 'a' + changes[i]) % 26 + 26) % 26;

      sb.append((char) (c + 'a'));
    }

    return sb.toString();
  }
}
